File: leetcode/33.cpp
// 二分搜索的变种。
class Solution {
public:
    int binary_search(const int ns[], const int from, const int to, const int target) {
        int mid = (from + to) / 2;
        
        if (from == to)
            return ns[from] == target ? from : -1;
        if (from > to)
            return -1;
        if (from + 1 == to) {
            if (ns[from] == target)
                return from;
            if (ns[to] == target)
                return to;
            else
                return -1;
        }
        
        if (ns[from] < ns[to]) {
            if (target < ns[mid]) {
                return binary_search(ns, from, mid - 1, target);
            } else if (target > ns[mid]) {
                return binary_search(ns, mid + 1, to, target);
            } else {
                return mid;
            }
        }
        
        else if (ns[from] < ns[mid]) {
            if (target == ns[mid]) {
                return mid;
            } else if (target >= ns[from] && target < ns[mid]) {
                return binary_search(ns, from, mid - 1, target);
            } else {
                return binary_search(ns, mid + 1, to, target);
            }
        }
        
        else {
            if (target == ns[mid]) {
                return mid;
            } else if (target > ns[mid] && target <= ns[to]) {
                return binary_search(ns, mid + 1, to, target);
            } else {
                return binary_search(ns, from, mid - 1, target);
            }
        }
    }

    int search(vector<int>& nums, int target) {
        int *ns = nums.data();
        int n = nums.size();
        return binary_search(ns, 0, n - 1, target);
    }
};

File: leetcode/33.rs
impl Solution {
    pub fn linear(a: &[i32], tg: i32, lb: usize) -> i32 {
        for i in 0..a.len() {
            if a[i] == tg {
                return (lb + i) as i32;
            }
        }
        return -1;
    }

    // assuming no duplicate
    pub fn search(a: Vec<i32>, tg: i32) -> i32 {
        let n = a.len();
        if n == 0 {
            return -1;
        }

        let (mut lb, mut ub) = (0, n - 1);
        loop {
            if ub - lb < 2 {
                return Self::linear(&a[lb..ub + 1], tg, lb);
            }

            let mid = (lb + ub) / 2;
            if a[mid] == tg {
                return mid as i32;
            }
            if a[lb] < a[ub] {
                if a[mid] < tg {
                    lb = mid + 1;
                } else {
                    ub = mid;
                }
            } else if a[lb] < a[mid] {
                if a[lb] <= tg && tg < a[mid] {
                    ub = mid;
                } else {
                    lb = mid + 1;
                }
            } else {
                if a[mid] < tg && tg <= a[ub] {
                    lb = mid + 1;
                } else {
                    ub = mid;
                }
            }
        }
    }
}
